src/components/Liveness/Recorder.js

```diff
@@ -0,0 +1,196 @@
+// @flow
+import * as React from 'react'
+import { h, Component } from 'preact'
+import classNames from 'classnames'
+import Challenge from './Challenge'
+import type { ChallengeType, ChallengeResultType } from './Challenge'
+import Camera from '../Camera'
+import CameraError from '../CameraError'
+import Timeout from '../Timeout'
+import Title from '../Title'
+import ToggleFullScreen from '../ToggleFullScreen'
+import { FaceOverlay } from '../Overlay'
+import theme from '../Theme/style.css'
+import style from './style.css'
+
+export type RecorderType = {
+  i18n: Object,
+  challenges: ChallengeType[],
+  onRedo: void => void,
+  onVideoRecorded: (?Blob, ?ChallengeResultType) => void,
+  onVideoRecordingStart: void => void,
+  onSwitchChallenge: void => void,
+  timeoutSeconds: number,
+  renderFallback: Function,
+  trackScreen: Function,
+  useFullScreen: Function,
+}
+
+type State = {
+  currentIndex: number,
+  isRecording: boolean,
+  hasBecomeInactive: boolean,
+  hasRecordingTakenTooLong: boolean,
+}
+
+const initialState = {
+  currentIndex: 0,
+  isRecording: false,
+  hasBecomeInactive: false,
+  hasRecordingTakenTooLong: false,
+}
+
+const inactiveError = { name: 'CAMERA_INACTIVE', type: 'warning' }
+const recordingTooLongError = { name: 'LIVENESS_TIMEOUT', type: 'warning' }
+
+export default class LivenessCamera extends Component<RecorderType, State> {
+  static defaultProps = {
+    timeoutSeconds: 20,
+  }
+
+  timeout: TimeoutID
+  webcam = null
+
+  state: State = { ...initialState }
+
+  componentWillReceiveProps(nextProps: RecorderType) {
+    if (nextProps.challenges !== this.props.challenges) {
+      this.setState({ ...initialState })
+    }
+  }
+
+  startRecording = () => {
+    this.webcam && this.webcam.startRecording()
+    this.setState({isRecording: true, hasBecomeInactive: false})
+  }
+
+  stopRecording = () => {
+    this.webcam && this.webcam.stopRecording()
+    this.setState({isRecording: false})
+  }
+
+  handleRecordingStart = () => {
+    this.startRecording()
+    this.props.onVideoRecordingStart()
+  }
+
+  handleRecordingStop = () => {
+    const { hasRecordingTakenTooLong } = this.state
+    this.stopRecording()
+    if (this.webcam && !hasRecordingTakenTooLong) {
+      this.props.onVideoRecorded(this.webcam.getVideoBlob())
+    }
+  }
+
+  handleNextChallenge = () => {
+    this.setState({ currentIndex: this.state.currentIndex + 1 })
+    this.props.onSwitchChallenge()
+  }
+
+  handleInactivityTimeout = () => {
+    this.setState({ hasBecomeInactive: true })
+  }
+
+  handleRecordingTimeout = () => {
+    this.setState({ hasRecordingTakenTooLong: true })
+    this.stopRecording()
+  }
+
+  redoActionsFallback = (text: string) =>
+    <span onClick={this.props.onRedo}>{text}</span>
+
+  renderError = () => {
```

Comment:
This `LivenessCamera` component is pretty big. Can we extract this into a pure component `RecorderError` ?